prov/efa/src/rdm/efa_rdm_ep_fiops.c

```diff
@@ -1288,6 +1289,60 @@ static int efa_rdm_ep_ctrl(struct fid *fid, int command, void *arg)
 	return ret;
 }
 
+static int efa_rdm_ep_cancel_match_recv(struct dlist_entry *item,
+				         const void *context)
+{
+	struct efa_rdm_ope *rxe = container_of(item, struct efa_rdm_ope, entry);
+	return rxe->cq_entry.op_context == context;
+}
+
+/**
+ * @brief Cancel a recv in ep->user_recv_rxe_list
+ *
+ * @param ep efa_rdm_ep
+ * @param context pointer to the context to be cancelled
+ * @return ssize_t 0 on success, negative integer on failure.
+ */
+static
+ssize_t efa_rdm_ep_cancel_zcpy_recv(struct efa_rdm_ep *ep,
+				     void *context)
+{
+	struct dlist_entry *entry;
+	struct efa_rdm_ope *rxe;
+	struct fi_cq_err_entry err_entry;
+	struct util_srx_ctx *srx_ctx;
+	int ret;
+
+	srx_ctx = efa_rdm_ep_get_peer_srx_ctx(ep);
+	ofi_genlock_lock(srx_ctx->lock);
+
+	entry = dlist_remove_first_match(&ep->user_recv_rxe_list,
+					 &efa_rdm_ep_cancel_match_recv,
+					 context);
+	if (!entry) {
+		ret = 0;
+		goto unlock;
+	}
+
+	rxe = container_of(entry, struct efa_rdm_ope, entry);
+	assert(rxe->user_rx_pkt);
+	rxe->user_rx_pkt->flags |= EFA_RDM_PKE_USER_RECV_CANCEL;
+
+	dlist_remove(&rxe->entry);
+	memset(&err_entry, 0, sizeof(err_entry));
+	err_entry.op_context = rxe->cq_entry.op_context;
+	err_entry.flags |= rxe->cq_entry.flags;
+	err_entry.err = FI_ECANCELED;
+	err_entry.prov_errno = -FI_ECANCELED;
```

Comment:
Shouldn't this be an EFA-specific code?

Reply:
recv cancel is a general libfabric operation, there is no efa specific thing at all

>fi_cancel attempts to cancel an outstanding asynchronous operation. Canceling an operation causes the fabric provider to search for the operation and, if it is still pending, complete it as having been canceled. An error queue entry will be available in the associated error queue with error code FI_ECANCELED.

Reply:
I don't think there is something else we can put on prov_errno, I am following https://github.com/ofiwg/libfabric/blob/main/prov/util/src/util_srx.c#L888-L890